{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Clair3 illumina demo",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true,
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/HKU-BAL/Clair3/blob/main/colab/clair3_illumina_quick_demo.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Y8OphKkrzRXq"
      },
      "source": [
        "<h1>Clair3 illumina demo</h1>\n",
        "\n",
        "The following is the document describing the calling workflow of `Clair3`. Clair3 is an open-source project available at https://github.com/HKU-BAL/Clair3.\n",
        "\n",
        "We will demonstrate Clair3 long-read variant calling workflow, including pileup calling and full-alignment calling.\n",
        "\n",
        "The document you are reading is not a static web page, but an interactive environment called a Colab notebook that lets you write and execute code. You can inspect, modify, and run any of the code on this page. \n",
        "\n",
        "***Please note that current workflow is using google colab resource, we only suggest using small datasets for benchmarking.***"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xrD1AW3-0O1w"
      },
      "source": [
        "## Installation\n",
        "\n",
        "Before getting started with how `Clair3` works, we will install it into the Colab environment. This will enable us to both inspect and run the code. To install the `Clair3`, run the code cell below by clicking the \"play\" icon to the left or pressing on the cell or type `<ctrl>-<enter>` to run the cell."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "24pW_7pEMoCE"
      },
      "source": [
        "### Build an anaconda virtual environment"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PJZWryaduqV5",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "73292d2e-590a-459f-f813-8f2fbf1de49d"
      },
      "source": [
        "# build an anaconda virtual environment\n",
        "!mkdir -p /content\n",
        "%cd /content\n",
        "!wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
        "!chmod +x ./Miniconda3-latest-Linux-x86_64.sh \n",
        "!bash ./Miniconda3-latest-Linux-x86_64.sh -f -b -p /usr/local"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/content\n",
            "--2021-06-16 12:08:14--  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.131.3, 104.16.130.3, 2606:4700::6810:8203, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.131.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 94235922 (90M) [application/x-sh]\n",
            "Saving to: ‘Miniconda3-latest-Linux-x86_64.sh.1’\n",
            "\n",
            "Miniconda3-latest-L 100%[===================>]  89.87M   123MB/s    in 0.7s    \n",
            "\n",
            "2021-06-16 12:08:14 (123 MB/s) - ‘Miniconda3-latest-Linux-x86_64.sh.1’ saved [94235922/94235922]\n",
            "\n",
            "PREFIX=/usr/local\n",
            "Unpacking payload ...\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\bdone\n",
            "Solving environment: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - _libgcc_mutex==0.1=main\n",
            "    - brotlipy==0.7.0=py38h27cfd23_1003\n",
            "    - ca-certificates==2020.10.14=0\n",
            "    - certifi==2020.6.20=pyhd3eb1b0_3\n",
            "    - cffi==1.14.3=py38h261ae71_2\n",
            "    - chardet==3.0.4=py38h06a4308_1003\n",
            "    - conda-package-handling==1.7.2=py38h03888b9_0\n",
            "    - conda==4.9.2=py38h06a4308_0\n",
            "    - cryptography==3.2.1=py38h3c74f83_1\n",
            "    - idna==2.10=py_0\n",
            "    - ld_impl_linux-64==2.33.1=h53a641e_7\n",
            "    - libedit==3.1.20191231=h14c3975_1\n",
            "    - libffi==3.3=he6710b0_2\n",
            "    - libgcc-ng==9.1.0=hdf63c60_0\n",
            "    - libstdcxx-ng==9.1.0=hdf63c60_0\n",
            "    - ncurses==6.2=he6710b0_1\n",
            "    - openssl==1.1.1h=h7b6447c_0\n",
            "    - pip==20.2.4=py38h06a4308_0\n",
            "    - pycosat==0.6.3=py38h7b6447c_1\n",
            "    - pycparser==2.20=py_2\n",
            "    - pyopenssl==19.1.0=pyhd3eb1b0_1\n",
            "    - pysocks==1.7.1=py38h06a4308_0\n",
            "    - python==3.8.5=h7579374_1\n",
            "    - readline==8.0=h7b6447c_0\n",
            "    - requests==2.24.0=py_0\n",
            "    - ruamel_yaml==0.15.87=py38h7b6447c_1\n",
            "    - setuptools==50.3.1=py38h06a4308_1\n",
            "    - six==1.15.0=py38h06a4308_0\n",
            "    - sqlite==3.33.0=h62c20be_0\n",
            "    - tk==8.6.10=hbc83047_0\n",
            "    - tqdm==4.51.0=pyhd3eb1b0_0\n",
            "    - urllib3==1.25.11=py_0\n",
            "    - wheel==0.35.1=pyhd3eb1b0_0\n",
            "    - xz==5.2.5=h7b6447c_0\n",
            "    - yaml==0.2.5=h7b6447c_0\n",
            "    - zlib==1.2.11=h7b6447c_3\n",
            "\n",
            "\n",
            "The following packages will be SUPERSEDED by a higher-priority channel:\n",
            "\n",
            "  ca-certificates    conda-forge::ca-certificates-2021.5.3~ --> pkgs/main::ca-certificates-2020.10.14-0\n",
            "  certifi            conda-forge/linux-64::certifi-2021.5.~ --> pkgs/main/noarch::certifi-2020.6.20-pyhd3eb1b0_3\n",
            "  conda              conda-forge::conda-4.10.1-py38h578d9b~ --> pkgs/main::conda-4.9.2-py38h06a4308_0\n",
            "\n",
            "The following packages will be DOWNGRADED:\n",
            "\n",
            "  openssl                                 1.1.1k-h27cfd23_0 --> 1.1.1h-h7b6447c_0\n",
            "\n",
            "\n",
            "Preparing transaction: | \b\bdone\n",
            "Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Miniconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Miniconda3: /usr/local\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "HyyI0VZ_zHd8"
      },
      "source": [
        "### Use conda environment to install python packages"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "LJDS3xRUu9A-",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "844f9cc0-d707-437c-807f-52f6bdb8242a"
      },
      "source": [
        "# create and activate an environment named clair3\n",
        "%%bash\n",
        "conda create -n clair3 python=3.6.10 -y\n",
        "source activate clair3\n",
        "\n",
        "conda install -c conda-forge pypy3.6 -y\n",
        "pypy3 -m ensurepip\n",
        "pypy3 -m pip install intervaltree==3.0.2\n",
        "pypy3 -m pip install mpmath==1.2.1\n",
        "\n",
        "# install python packages in environment\n",
        "pip3 install tensorflow==2.2.0\n",
        "pip3 install intervaltree==3.0.2  tensorflow-addons==0.11.2 tables==3.6.1\n",
        "conda install -c anaconda pigz==2.4 -y\n",
        "conda install -c conda-forge parallel=20191122 zstd=1.4.4 -y\n",
        "conda install -c conda-forge -c bioconda samtools=1.10 -y\n",
        "conda install -c conda-forge -c bioconda whatshap=1.0 -y\n",
        "\n",
        "cd /content\n",
        "git clone https://github.com/HKU-BAL/Clair3.git\n",
        "cd /content/Clair3\n",
        "# download pre-trained models\n",
        "mkdir models\n",
        "wget --quiet http://www.bio8.cs.hku.hk/clair3/clair3_models/clair3_models.tar.gz \n",
        "tar -zxvf clair3_models.tar.gz -C ./models\n",
        "\n",
        "# need to install boost library for illumina platform\n",
        "conda install -c conda-forge boost=1.67.0 -y\n",
        "echo \"Environment:\" ${CONDA_PREFIX}\n",
        "cd /content/Clair3/preprocess/realign\n",
        "g++ -std=c++14 -O1 -shared -fPIC -o realigner ssw_cpp.cpp ssw.c realigner.cpp\n",
        "g++ -std=c++11 -shared -fPIC -o debruijn_graph -O3 debruijn_graph.cpp -I ${CONDA_PREFIX}/include -L ${CONDA_PREFIX}/lib"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - python=3.6.10\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    pip-21.1.2                 |     pyhd8ed1ab_0         1.1 MB  conda-forge\n",
            "    setuptools-49.6.0          |   py36h5fab9bb_3         936 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         2.0 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
            "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
            "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.5.30-ha878542_0\n",
            "  certifi            conda-forge/linux-64::certifi-2021.5.30-py36h5fab9bb_0\n",
            "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.35.1-hea4e1c9_2\n",
            "  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007\n",
            "  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.3.0-h2828fa1_19\n",
            "  libgomp            conda-forge/linux-64::libgomp-9.3.0-h2828fa1_19\n",
            "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.3.0-h6de172a_19\n",
            "  ncurses            conda-forge/linux-64::ncurses-6.2-h58526e2_4\n",
            "  openssl            conda-forge/linux-64::openssl-1.1.1k-h7f98852_0\n",
            "  pip                conda-forge/noarch::pip-21.1.2-pyhd8ed1ab_0\n",
            "  python             conda-forge/linux-64::python-3.6.10-h8356626_1011_cpython\n",
            "  python_abi         conda-forge/linux-64::python_abi-3.6-1_cp36m\n",
            "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
            "  setuptools         conda-forge/linux-64::setuptools-49.6.0-py36h5fab9bb_3\n",
            "  sqlite             conda-forge/linux-64::sqlite-3.35.5-h74cdb3f_0\n",
            "  tk                 conda-forge/linux-64::tk-8.6.10-h21135ba_1\n",
            "  wheel              conda-forge/noarch::wheel-0.36.2-pyhd3deb0d_0\n",
            "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
            "  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1010\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "\rpip-21.1.2           | 1.1 MB    |            |   0% \rpip-21.1.2           | 1.1 MB    | #2         |  13% \rpip-21.1.2           | 1.1 MB    | ########## | 100% \rpip-21.1.2           | 1.1 MB    | ########## | 100% \n",
            "\rsetuptools-49.6.0    | 936 KB    |            |   0% \rsetuptools-49.6.0    | 936 KB    | ########## | 100% \rsetuptools-49.6.0    | 936 KB    | ########## | 100% \n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate clair3\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.\n",
            "Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - pypy3.6\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
            "  expat              conda-forge/linux-64::expat-2.4.1-h9c3ff4c_0\n",
            "  gdbm               conda-forge/linux-64::gdbm-1.18-h0a1914f_2\n",
            "  pypy3.6            conda-forge/linux-64::pypy3.6-7.3.2-h45e8706_2\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "Looking in links: /tmp/tmpekltudtr\n",
            "Requirement already satisfied: setuptools in /usr/local/envs/clair3/lib/python3.6/site-packages (49.6.0.post20210108)\n",
            "Requirement already satisfied: pip in /usr/local/envs/clair3/lib/python3.6/site-packages (21.1.2)\n",
            "Collecting intervaltree==3.0.2\n",
            "  Using cached intervaltree-3.0.2-py3-none-any.whl\n",
            "Collecting sortedcontainers<3.0,>=2.0\n",
            "  Using cached sortedcontainers-2.4.0-py2.py3-none-any.whl (29 kB)\n",
            "Installing collected packages: sortedcontainers, intervaltree\n",
            "Successfully installed intervaltree-3.0.2 sortedcontainers-2.4.0\n",
            "Collecting mpmath==1.2.1\n",
            "  Using cached mpmath-1.2.1-py3-none-any.whl (532 kB)\n",
            "Installing collected packages: mpmath\n",
            "Successfully installed mpmath-1.2.1\n",
            "Collecting tensorflow==2.2.0\n",
            "  Using cached tensorflow-2.2.0-cp36-cp36m-manylinux2010_x86_64.whl (516.2 MB)\n",
            "Collecting opt-einsum>=2.3.2\n",
            "  Using cached opt_einsum-3.3.0-py3-none-any.whl (65 kB)\n",
            "Collecting google-pasta>=0.1.8\n",
            "  Using cached google_pasta-0.2.0-py3-none-any.whl (57 kB)\n",
            "Collecting tensorflow-estimator<2.3.0,>=2.2.0\n",
            "  Using cached tensorflow_estimator-2.2.0-py2.py3-none-any.whl (454 kB)\n",
            "Requirement already satisfied: wheel>=0.26 in /usr/local/envs/clair3/lib/python3.6/site-packages (from tensorflow==2.2.0) (0.36.2)\n",
            "Collecting termcolor>=1.1.0\n",
            "  Using cached termcolor-1.1.0-py3-none-any.whl\n",
            "Collecting numpy<2.0,>=1.16.0\n",
            "  Using cached numpy-1.19.5-cp36-cp36m-manylinux2010_x86_64.whl (14.8 MB)\n",
            "Collecting protobuf>=3.8.0\n",
            "  Using cached protobuf-3.17.3-cp36-cp36m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (1.0 MB)\n",
            "Collecting gast==0.3.3\n",
            "  Using cached gast-0.3.3-py2.py3-none-any.whl (9.7 kB)\n",
            "Collecting astunparse==1.6.3\n",
            "  Using cached astunparse-1.6.3-py2.py3-none-any.whl (12 kB)\n",
            "Collecting h5py<2.11.0,>=2.10.0\n",
            "  Using cached h5py-2.10.0-cp36-cp36m-manylinux1_x86_64.whl (2.9 MB)\n",
            "Collecting keras-preprocessing>=1.1.0\n",
            "  Using cached Keras_Preprocessing-1.1.2-py2.py3-none-any.whl (42 kB)\n",
            "Collecting scipy==1.4.1\n",
            "  Using cached scipy-1.4.1-cp36-cp36m-manylinux1_x86_64.whl (26.1 MB)\n",
            "Collecting grpcio>=1.8.6\n",
            "  Using cached grpcio-1.38.0-cp36-cp36m-manylinux2014_x86_64.whl (4.2 MB)\n",
            "Collecting six>=1.12.0\n",
            "  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)\n",
            "Collecting tensorboard<2.3.0,>=2.2.0\n",
            "  Using cached tensorboard-2.2.2-py3-none-any.whl (3.0 MB)\n",
            "Collecting absl-py>=0.7.0\n",
            "  Using cached absl_py-0.13.0-py3-none-any.whl (132 kB)\n",
            "Collecting wrapt>=1.11.1\n",
            "  Using cached wrapt-1.12.1-cp36-cp36m-linux_x86_64.whl\n",
            "Collecting requests<3,>=2.21.0\n",
            "  Using cached requests-2.25.1-py2.py3-none-any.whl (61 kB)\n",
            "Collecting google-auth<2,>=1.6.3\n",
            "  Using cached google_auth-1.31.0-py2.py3-none-any.whl (147 kB)\n",
            "Collecting werkzeug>=0.11.15\n",
            "  Using cached Werkzeug-2.0.1-py3-none-any.whl (288 kB)\n",
            "Collecting markdown>=2.6.8\n",
            "  Using cached Markdown-3.3.4-py3-none-any.whl (97 kB)\n",
            "Requirement already satisfied: setuptools>=41.0.0 in /usr/local/envs/clair3/lib/python3.6/site-packages (from tensorboard<2.3.0,>=2.2.0->tensorflow==2.2.0) (49.6.0.post20210108)\n",
            "Collecting google-auth-oauthlib<0.5,>=0.4.1\n",
            "  Using cached google_auth_oauthlib-0.4.4-py2.py3-none-any.whl (18 kB)\n",
            "Collecting tensorboard-plugin-wit>=1.6.0\n",
            "  Using cached tensorboard_plugin_wit-1.8.0-py3-none-any.whl (781 kB)\n",
            "Collecting cachetools<5.0,>=2.0.0\n",
            "  Using cached cachetools-4.2.2-py3-none-any.whl (11 kB)\n",
            "Collecting rsa<5,>=3.1.4\n",
            "  Using cached rsa-4.7.2-py3-none-any.whl (34 kB)\n",
            "Collecting pyasn1-modules>=0.2.1\n",
            "  Using cached pyasn1_modules-0.2.8-py2.py3-none-any.whl (155 kB)\n",
            "Collecting requests-oauthlib>=0.7.0\n",
            "  Using cached requests_oauthlib-1.3.0-py2.py3-none-any.whl (23 kB)\n",
            "Collecting importlib-metadata\n",
            "  Using cached importlib_metadata-4.5.0-py3-none-any.whl (17 kB)\n",
            "Collecting pyasn1<0.5.0,>=0.4.6\n",
            "  Using cached pyasn1-0.4.8-py2.py3-none-any.whl (77 kB)\n",
            "Collecting chardet<5,>=3.0.2\n",
            "  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)\n",
            "Collecting urllib3<1.27,>=1.21.1\n",
            "  Using cached urllib3-1.26.5-py2.py3-none-any.whl (138 kB)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/envs/clair3/lib/python3.6/site-packages (from requests<3,>=2.21.0->tensorboard<2.3.0,>=2.2.0->tensorflow==2.2.0) (2021.5.30)\n",
            "Collecting idna<3,>=2.5\n",
            "  Using cached idna-2.10-py2.py3-none-any.whl (58 kB)\n",
            "Collecting oauthlib>=3.0.0\n",
            "  Using cached oauthlib-3.1.1-py2.py3-none-any.whl (146 kB)\n",
            "Collecting dataclasses\n",
            "  Using cached dataclasses-0.8-py3-none-any.whl (19 kB)\n",
            "Collecting typing-extensions>=3.6.4\n",
            "  Using cached typing_extensions-3.10.0.0-py3-none-any.whl (26 kB)\n",
            "Collecting zipp>=0.5\n",
            "  Using cached zipp-3.4.1-py3-none-any.whl (5.2 kB)\n",
            "Installing collected packages: urllib3, pyasn1, idna, chardet, zipp, typing-extensions, six, rsa, requests, pyasn1-modules, oauthlib, cachetools, requests-oauthlib, importlib-metadata, google-auth, dataclasses, werkzeug, tensorboard-plugin-wit, protobuf, numpy, markdown, grpcio, google-auth-oauthlib, absl-py, wrapt, termcolor, tensorflow-estimator, tensorboard, scipy, opt-einsum, keras-preprocessing, h5py, google-pasta, gast, astunparse, tensorflow\n",
            "Successfully installed absl-py-0.13.0 astunparse-1.6.3 cachetools-4.2.2 chardet-4.0.0 dataclasses-0.8 gast-0.3.3 google-auth-1.31.0 google-auth-oauthlib-0.4.4 google-pasta-0.2.0 grpcio-1.38.0 h5py-2.10.0 idna-2.10 importlib-metadata-4.5.0 keras-preprocessing-1.1.2 markdown-3.3.4 numpy-1.19.5 oauthlib-3.1.1 opt-einsum-3.3.0 protobuf-3.17.3 pyasn1-0.4.8 pyasn1-modules-0.2.8 requests-2.25.1 requests-oauthlib-1.3.0 rsa-4.7.2 scipy-1.4.1 six-1.16.0 tensorboard-2.2.2 tensorboard-plugin-wit-1.8.0 tensorflow-2.2.0 tensorflow-estimator-2.2.0 termcolor-1.1.0 typing-extensions-3.10.0.0 urllib3-1.26.5 werkzeug-2.0.1 wrapt-1.12.1 zipp-3.4.1\n",
            "Requirement already satisfied: intervaltree==3.0.2 in /usr/local/envs/clair3/lib/python3.6/site-packages (3.0.2)\n",
            "Collecting tensorflow-addons==0.11.2\n",
            "  Using cached tensorflow_addons-0.11.2-cp36-cp36m-manylinux2010_x86_64.whl (1.1 MB)\n",
            "Collecting tables==3.6.1\n",
            "  Using cached tables-3.6.1-cp36-cp36m-manylinux1_x86_64.whl (4.3 MB)\n",
            "Requirement already satisfied: sortedcontainers<3.0,>=2.0 in /usr/local/envs/clair3/lib/python3.6/site-packages (from intervaltree==3.0.2) (2.4.0)\n",
            "Collecting typeguard>=2.7\n",
            "  Using cached typeguard-2.12.1-py3-none-any.whl (17 kB)\n",
            "Collecting numexpr>=2.6.2\n",
            "  Using cached numexpr-2.7.3-cp36-cp36m-manylinux2010_x86_64.whl (469 kB)\n",
            "Requirement already satisfied: numpy>=1.9.3 in /usr/local/envs/clair3/lib/python3.6/site-packages (from tables==3.6.1) (1.19.5)\n",
            "Installing collected packages: typeguard, numexpr, tensorflow-addons, tables\n",
            "Successfully installed numexpr-2.7.3 tables-3.6.1 tensorflow-addons-0.11.2 typeguard-2.12.1\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - pigz==2.4\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    ca-certificates-2020.10.14 |                0         128 KB  anaconda\n",
            "    openssl-1.1.1h             |       h7b6447c_0         3.8 MB  anaconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:         3.9 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  pigz               anaconda/linux-64::pigz-2.4-h84994c4_0\n",
            "\n",
            "The following packages will be SUPERSEDED by a higher-priority channel:\n",
            "\n",
            "  ca-certificates    conda-forge::ca-certificates-2021.5.3~ --> anaconda::ca-certificates-2020.10.14-0\n",
            "  certifi            conda-forge::certifi-2021.5.30-py36h5~ --> anaconda::certifi-2020.6.20-py36_0\n",
            "  openssl            conda-forge::openssl-1.1.1k-h7f98852_0 --> anaconda::openssl-1.1.1h-h7b6447c_0\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "\rca-certificates-2020 | 128 KB    |            |   0% \rca-certificates-2020 | 128 KB    | #####      |  50% \rca-certificates-2020 | 128 KB    | ########## | 100% \n",
            "\ropenssl-1.1.1h       | 3.8 MB    |            |   0% \ropenssl-1.1.1h       | 3.8 MB    | ########## | 100% \ropenssl-1.1.1h       | 3.8 MB    | ########## | 100% \n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - parallel=20191122\n",
            "    - zstd=1.4.4\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  lz4-c              conda-forge/linux-64::lz4-c-1.9.2-he1b5a44_3\n",
            "  parallel           conda-forge/linux-64::parallel-20191122-0\n",
            "  perl               conda-forge/linux-64::perl-5.32.1-0_h7f98852_perl5\n",
            "  zstd               conda-forge/linux-64::zstd-1.4.4-h6597ccf_3\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  ca-certificates    anaconda::ca-certificates-2020.10.14-0 --> conda-forge::ca-certificates-2021.5.30-ha878542_0\n",
            "  certifi                anaconda::certifi-2020.6.20-py36_0 --> conda-forge::certifi-2021.5.30-py36h5fab9bb_0\n",
            "  openssl               anaconda::openssl-1.1.1h-h7b6447c_0 --> conda-forge::openssl-1.1.1k-h7f98852_0\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - samtools=1.10\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  c-ares             conda-forge/linux-64::c-ares-1.17.1-h7f98852_1\n",
            "  htslib             bioconda/linux-64::htslib-1.10.2-hd3b49d5_1\n",
            "  krb5               conda-forge/linux-64::krb5-1.19.1-hcc1bbae_0\n",
            "  libcurl            conda-forge/linux-64::libcurl-7.77.0-h2574ce0_0\n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.6-h516909a_0\n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
            "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
            "  libnghttp2         conda-forge/linux-64::libnghttp2-1.43.0-h812cca2_0\n",
            "  libssh2            conda-forge/linux-64::libssh2-1.9.0-ha56f1ee_6\n",
            "  samtools           bioconda/linux-64::samtools-1.10-h2e538c0_3\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with initial frozen solve. Retrying with flexible solve.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - whatshap=1.0\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  biopython          conda-forge/linux-64::biopython-1.79-py36h8f6f2f9_0\n",
            "  decorator          conda-forge/noarch::decorator-5.0.9-pyhd8ed1ab_0\n",
            "  isa-l              conda-forge/linux-64::isa-l-2.30.0-ha770c72_4\n",
            "  libblas            conda-forge/linux-64::libblas-3.9.0-9_openblas\n",
            "  libcblas           conda-forge/linux-64::libcblas-3.9.0-9_openblas\n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-9.3.0-hff62375_19\n",
            "  libgfortran5       conda-forge/linux-64::libgfortran5-9.3.0-hff62375_19\n",
            "  liblapack          conda-forge/linux-64::liblapack-3.9.0-9_openblas\n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.15-pthreads_h8fe5266_1\n",
            "  networkx           conda-forge/noarch::networkx-2.5-py_0\n",
            "  numpy              conda-forge/linux-64::numpy-1.19.5-py36h2aa4a07_1\n",
            "  pyfaidx            bioconda/noarch::pyfaidx-0.5.9.5-pyh3252c3a_0\n",
            "  pysam              bioconda/linux-64::pysam-0.16.0.1-py36h4c34d4e_1\n",
            "  python-isal        conda-forge/linux-64::python-isal-0.10.0-py36h8f6f2f9_0\n",
            "  scipy              conda-forge/linux-64::scipy-1.5.3-py36h9e8f40b_0\n",
            "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0\n",
            "  whatshap           bioconda/linux-64::whatshap-1.0-py36hf1ae8f4_1\n",
            "  xopen              conda-forge/linux-64::xopen-1.1.0-py36h5fab9bb_2\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "ilmn/\n",
            "ilmn/full_alignment.data-00000-of-00002\n",
            "ilmn/full_alignment.data-00001-of-00002\n",
            "ilmn/pileup.data-00001-of-00002\n",
            "ilmn/pileup.data-00000-of-00002\n",
            "ilmn/full_alignment.index\n",
            "ilmn/pileup.index\n",
            "hifi/\n",
            "hifi/pileup.data-00000-of-00002\n",
            "hifi/pileup.index\n",
            "hifi/pileup.data-00001-of-00002\n",
            "hifi/full_alignment.data-00000-of-00002\n",
            "hifi/full_alignment.data-00001-of-00002\n",
            "hifi/full_alignment.index\n",
            "ont/\n",
            "ont/full_alignment.data-00000-of-00002\n",
            "ont/pileup.index\n",
            "ont/full_alignment.data-00001-of-00002\n",
            "ont/pileup.data-00000-of-00002\n",
            "ont/full_alignment.index\n",
            "ont/pileup.data-00001-of-00002\n",
            "ont_guppy5/\n",
            "ont_guppy5/full_alignment.data-00001-of-00002\n",
            "ont_guppy5/pileup.index\n",
            "ont_guppy5/full_alignment.data-00000-of-00002\n",
            "ont_guppy5/pileup.data-00001-of-00002\n",
            "ont_guppy5/pileup.data-00000-of-00002\n",
            "ont_guppy5/full_alignment.index\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/clair3\n",
            "\n",
            "  added / updated specs:\n",
            "    - boost=1.67.0\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  boost              conda-forge/linux-64::boost-1.67.0-py36h3e44d54_0\n",
            "  boost-cpp          conda-forge/linux-64::boost-cpp-1.67.0-h3a22d5f_0\n",
            "  icu                conda-forge/linux-64::icu-58.2-hf484d3e_1000\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "Environment: /usr/local/envs/clair3\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "WARNING: Running pip as root will break packages and permissions. You should install packages reliably by using venv: https://pip.pypa.io/warnings/venv\n",
            "WARNING: Running pip as root will break packages and permissions. You should install packages reliably by using venv: https://pip.pypa.io/warnings/venv\n",
            "WARNING: Running pip as root will break packages and permissions. You should install packages reliably by using venv: https://pip.pypa.io/warnings/venv\n",
            "WARNING: Running pip as root will break packages and permissions. You should install packages reliably by using venv: https://pip.pypa.io/warnings/venv\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "fatal: destination path 'Clair3' already exists and is not an empty directory.\n",
            "mkdir: cannot create directory ‘models’: File exists\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "DhwqwvbD1e5s"
      },
      "source": [
        "## Illumina quick demo\n",
        "\n",
        "Here is a quick demo for the Illumina variant calling using GIAB HG003 chromosome 20 data. We select a small chunked regions (chr20:100000-300000) to run Clair3 workflow."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "xiEyr06dzFmH",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "81397d1b-aeb0-4508-af73-562e3a9964b4"
      },
      "source": [
        "%%bash\n",
        "source activate clair3\n",
        "PLATFORM=\"ilmn\"\n",
        "INPUT_DIR=\"/content/data\"\n",
        "OUTPUT_DIR=\"/content/clair3_ilmn_demo\"\n",
        "THREADS=4        \n",
        "mkdir -p ${INPUT_DIR}\n",
        "mkdir -p ${OUTPUT_DIR}\n",
        "\n",
        "# download quick demo data\n",
        "# GRCh38_no_alt Reference\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/GRCh38_chr20.fa\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/GRCh38_chr20.fa.fai\n",
        "# BAM chr20:100000-300000\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/HG003_chr20_demo.bam\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/HG003_chr20_demo.bam.bai\n",
        "# GIAB Truth VCF and BED\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/HG003_GRCh38_chr20_v4.2.1_benchmark.vcf.gz\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/HG003_GRCh38_chr20_v4.2.1_benchmark.vcf.gz.tbi\n",
        "wget --quiet -P ${INPUT_DIR} http://www.bio8.cs.hku.hk/clair3/demo/quick_demo/illumina/HG003_GRCh38_chr20_v4.2.1_benchmark_noinconsistent.bed\n",
        "\n",
        "REF=\"GRCh38_chr20.fa\"\n",
        "BAM=\"HG003_chr20_demo.bam\"\n",
        "\n",
        "CONTIGS=\"chr20\"\n",
        "START_POS=100000\n",
        "END_POS=300000\n",
        "echo -e \"${CONTIGS}\\t${START_POS}\\t${END_POS}\" > ${INPUT_DIR}/quick_demo.bed\n",
        "\n",
        "# run Clair3 using one command\n",
        "cd /content/Clair3\n",
        "./run_clair3.sh \\\n",
        "  --bam_fn=${INPUT_DIR}/${BAM} \\\n",
        "  --ref_fn=${INPUT_DIR}/${REF} \\\n",
        "  --threads=${THREADS} \\\n",
        "  --platform=${PLATFORM} \\\n",
        "  --model_path=`pwd`\"/models/${PLATFORM}\" \\\n",
        "  --output=${OUTPUT_DIR} \\\n",
        "  --bed_fn=${INPUT_DIR}/quick_demo.bed\n"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[INFO] BAM FILE PATH: /content/data/HG003_chr20_demo.bam\n",
            "[INFO] REFERENCE FILE PATH: /content/data/GRCh38_chr20.fa\n",
            "[INFO] MODEL PATH: /content/Clair3/models/ilmn\n",
            "[INFO] OUTPUT FOLDER: /content/clair3_ilmn_demo\n",
            "[INFO] PLATFORM: ilmn\n",
            "[INFO] THREADS: 4\n",
            "[INFO] BED FILE PATH: /content/data/quick_demo.bed\n",
            "[INFO] VCF FILE PATH: EMPTY\n",
            "[INFO] CONTIGS: EMPTY\n",
            "[INFO] SAMTOOLS PATH: samtools\n",
            "[INFO] PYTHON PATH: python3\n",
            "[INFO] PYPY PATH: pypy3\n",
            "[INFO] PARALLEL PATH: parallel\n",
            "[INFO] WHATSHAP PATH: whatshap\n",
            "[INFO] CHUNK SIZE: 5000000\n",
            "[INFO] CHUNK NUM: 0\n",
            "[INFO] FULL ALIGN PROPORTION: 0.3\n",
            "[INFO] FULL ALIGN RERFERENCE PROPORTION: 0.3\n",
            "[INFO] USER DEFINED SNP THRESHOLD: 0.0\n",
            "[INFO] USER DEFINED INDEL THRESHOLD: 0.0\n",
            "[INFO] ENABLE FILEUP ONLY CALLING: False\n",
            "[INFO] ENABLE FAST MODE CALLING: False\n",
            "[INFO] ENABLE PRINTING REFERENCE CALLS: False\n",
            "[INFO] ENABLE OUTPUT GVCF: False\n",
            "[INFO] ENABLE HAPLOID PRECISE MODE: False\n",
            "[INFO] ENABLE HAPLOID SENSITIVE MODE: False\n",
            "[INFO] ENABLE INCLUDE ALL CTGS CALLING: False\n",
            "[INFO] ENABLE NO PHASING FOR FULL ALIGNMENT: False\n",
            "\n",
            "\u001b[33m[WARNING] Threads setting exceeds maximum available threads 2, set threads=2\u001b[0m\n",
            "+ ./scripts/clair3.sh --bam_fn /content/data/HG003_chr20_demo.bam --ref_fn /content/data/GRCh38_chr20.fa --threads 2 --model_path /content/Clair3/models/ilmn --platform ilmn --output /content/clair3_ilmn_demo --bed_fn=/content/data/quick_demo.bed --vcf_fn=EMPTY --ctg_name=EMPTY --sample_name=SAMPLE --chunk_num=0 --chunk_size=5000000 --samtools=samtools --python=python3 --pypy=pypy3 --parallel=parallel --whatshap=whatshap --qual=2 --var_pct_full=0.3 --ref_pct_full=0.3 --snp_min_af=0.0 --indel_min_af=0.0 --pileup_only=False --gvcf=False --fast_mode=False --print_ref_calls=False --haploid_precise=False --haploid_sensitive=False --include_all_ctgs=False --no_phasing_for_fa=False --pileup_model_prefix=pileup --fa_model_prefix=full_alignment\n",
            "[INFO] Check envrionment variables\n",
            "[INFO] --include_all_ctgs not enabled, use chr{1..22,X,Y} and {1..22,X,Y} by default\n",
            "[INFO] Call variant in contigs: chr20\n",
            "[INFO] Chunk number for each contig: 13\n",
            "[INFO] 1/7 Calling variants using pileup model\n",
            "[INFO] Delay 8 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 : 37\n",
            "Total time elapsed: 3.09 s\n",
            "[INFO] Delay 7 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 1/13 : 46\n",
            "Total time elapsed: 3.01 s\n",
            "[INFO] Delay 3 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 2/13 : 61\n",
            "Total time elapsed: 3.16 s\n",
            "[INFO] Delay 4 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 3/13 : 119\n",
            "Total time elapsed: 3.31 s\n",
            "[INFO] Delay 8 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 4/13 : 92\n",
            "Total time elapsed: 3.28 s\n",
            "[INFO] Delay 4 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 5/13 : 43\n",
            "Total time elapsed: 3.10 s\n",
            "[INFO] Delay 3 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 6/13 : 54\n",
            "Total time elapsed: 3.14 s\n",
            "[INFO] Delay 8 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 7/13 : 41\n",
            "Total time elapsed: 3.01 s\n",
            "[INFO] Delay 0 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 8/13 : 54\n",
            "Total time elapsed: 3.11 s\n",
            "[INFO] Delay 2 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 9/13 : 73\n",
            "Total time elapsed: 3.28 s\n",
            "[INFO] Delay 6 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 10/13 : 72\n",
            "Total time elapsed: 3.23 s\n",
            "[INFO] Delay 2 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 11/13 : 98\n",
            "Total time elapsed: 3.18 s\n",
            "[INFO] Delay 5 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 with chunk 12/13 : 72\n",
            "Total time elapsed: 3.29 s\n",
            "\n",
            "real\t2m25.251s\n",
            "user\t1m24.615s\n",
            "sys\t0m7.138s\n",
            "[INFO] Merge chunked contigs vcf files\n",
            "[INFO] 2/7 Filter Hete SNP varaints for Whatshap phasing and haplotag\n",
            "[INFO] Select phasing quality cut off 21\n",
            "[INFO] Total hete snp positions pass filtering: chr20: 254\n",
            "\n",
            "real\t0m0.290s\n",
            "user\t0m0.209s\n",
            "sys\t0m0.089s\n",
            "[INFO] 3/7 Whatshap phase vcf file\n",
            "This is WhatsHap 1.0 running under Python 3.6.10\n",
            "Working on 1 samples from 1 family\n",
            "======== Working on chromosome 'chr20'\n",
            "---- Processing individual SAMPLE\n",
            "Using maximum coverage per sample of 15X\n",
            "Number of variants skipped due to missing genotypes: 0\n",
            "Number of remaining heterozygous variants: 254\n",
            "Reading alignments and detecting alleles ...\n",
            "Found 7199 reads covering 254 variants\n",
            "Kept 2190 reads that cover at least two variants each\n",
            "Reducing coverage to at most 15X by selecting most informative reads ...\n",
            "Selected 954 reads covering 222 variants\n",
            "Variants covered by at least one phase-informative read in at least one individual after read selection: 222\n",
            "Phasing 1 sample by solving the MEC problem ...\n",
            "MEC cost: 210\n",
            "No. of phased blocks: 44\n",
            "Largest component contains 29 variants (13.1% of accessible variants) between position 155384 and 160212\n",
            "======== Writing VCF\n",
            "Done writing VCF\n",
            "\n",
            "== SUMMARY ==\n",
            "Maximum memory usage: 0.215 GB\n",
            "Time spent reading BAM/CRAM:                    1.9 s\n",
            "Time spent parsing VCF:                         0.0 s\n",
            "Time spent selecting reads:                     0.1 s\n",
            "Time spent phasing:                             0.8 s\n",
            "Time spent writing VCF:                         0.0 s\n",
            "Time spent finding components:                  0.0 s\n",
            "Time spent on rest:                             0.0 s\n",
            "Total elapsed time:                             2.8 s\n",
            "\n",
            "real\t0m3.737s\n",
            "user\t0m2.357s\n",
            "sys\t0m0.324s\n",
            "[INFO] 4/7 Whatshap haplotag input bam file\n",
            "Found 1 samples in input VCF\n",
            "Keeping 1 samples for haplo-tagging\n",
            "Found 0 samples in BAM file\n",
            "Reading alignments and detecting alleles ...\n",
            "Found 5992 reads covering 222 variants\n",
            "\n",
            "== SUMMARY ==\n",
            "Total alignments processed:                     62885\n",
            "Alignments that could be tagged:                11916\n",
            "Alignments spanning multiple phase sets:            0\n",
            "haplotag - total processing time: 2.3015453815460205\n",
            "\n",
            "real\t0m3.162s\n",
            "user\t0m2.371s\n",
            "sys\t0m0.777s\n",
            "\n",
            "real\t0m0.169s\n",
            "user\t0m0.164s\n",
            "sys\t0m0.040s\n",
            "[INFO] 5/7 Select candidates for full alignment\n",
            "[INFO] Select variant quality cut off 21.0\n",
            "[INFO] Select reference quality cut off 22.0\n",
            "[INFO] Total low qual ref variants to procceed in chr20: 94\n",
            "[INFO] Total low qual variants to procceed in chr20: 160\n",
            "\n",
            "real\t0m0.295s\n",
            "user\t0m0.225s\n",
            "sys\t0m0.087s\n",
            "[INFO] 6/7 Calling variants using Full Alignment\n",
            "[INFO] Delay 2 seconds before starting variant calling ...\n",
            "[mpileup] 1 samples in 1 input files\n",
            "Calling variants ...\n",
            "Total process positions in chr20 : 846\n",
            "Total time elapsed: 28.31 s\n",
            "\n",
            "real\t0m33.876s\n",
            "user\t0m39.636s\n",
            "sys\t0m1.856s\n",
            "[INFO] 7/7 Merge pileup vcf and full alignment vcf\n",
            "Merge pileup and full alignment VCF/GVCF\n",
            "[INFO] Pileup positions variants proceeded in chr20: 549\n",
            "[INFO] Realigned positions variants proceeded in chr20: 719\n",
            "\n",
            "real\t0m0.497s\n",
            "user\t0m0.379s\n",
            "sys\t0m0.136s\n",
            "[INFO] Finish calling, output file: /content/clair3_ilmn_demo/merge_output.vcf.gz\n",
            "\n",
            "real\t3m10.610s\n",
            "user\t2m11.519s\n",
            "sys\t0m10.941s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MrASkywtm17i"
      },
      "source": [
        "## Benchmarking\n",
        "\n",
        "We used `hap.py` version 0.3.12 to assess the variant calls against GIAB truth set. The `hap.py`\n",
        "program can be installed using conda. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4kI1wIMPi2sx",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9435814a-2f28-463c-d4e2-de50fd5690e4"
      },
      "source": [
        "# install hap.py without root privileges using conda\n",
        "%%bash\n",
        "conda config --add channels defaults\n",
        "conda config --add channels bioconda\n",
        "conda config --add channels conda-forge\n",
        "conda create -n happy-env -c bioconda hap.py -y\n",
        "conda install -c bioconda rtg-tools -y\n",
        "\n"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... failed with repodata from current_repodata.json, will retry with next repodata source.\n",
            "Collecting package metadata (repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/happy-env\n",
            "\n",
            "  added / updated specs:\n",
            "    - hap.py\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
            "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
            "  bcftools           bioconda/linux-64::bcftools-1.10.2-h4f4756c_3\n",
            "  bedtools           bioconda/linux-64::bedtools-2.30.0-h7d7f7ad_1\n",
            "  bx-python          bioconda/linux-64::bx-python-0.8.9-py27h54ae540_2\n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
            "  ca-certificates    conda-forge/linux-64::ca-certificates-2021.5.30-ha878542_0\n",
            "  certifi            conda-forge/linux-64::certifi-2019.11.28-py27h8c360ce_1\n",
            "  gsl                conda-forge/linux-64::gsl-2.6-he838d99_2\n",
            "  hap.py             bioconda/linux-64::hap.py-0.3.14-py27h5c5a3ab_0\n",
            "  htslib             bioconda/linux-64::htslib-1.10.2-hd3b49d5_1\n",
            "  krb5               conda-forge/linux-64::krb5-1.17.2-h926e7f8_0\n",
            "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.35.1-hea4e1c9_2\n",
            "  libblas            conda-forge/linux-64::libblas-3.9.0-8_openblas\n",
            "  libcblas           conda-forge/linux-64::libcblas-3.9.0-8_openblas\n",
            "  libcurl            conda-forge/linux-64::libcurl-7.71.1-hcdd3856_3\n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.6-h516909a_0\n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
            "  libffi             conda-forge/linux-64::libffi-3.2.1-he1b5a44_1007\n",
            "  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.3.0-h2828fa1_19\n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-7.5.0-h14aa051_19\n",
            "  libgfortran4       conda-forge/linux-64::libgfortran4-7.5.0-h14aa051_19\n",
            "  libgomp            conda-forge/linux-64::libgomp-9.3.0-h2828fa1_19\n",
            "  liblapack          conda-forge/linux-64::liblapack-3.9.0-8_openblas\n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.12-pthreads_hb3c22a3_1\n",
            "  libssh2            conda-forge/linux-64::libssh2-1.9.0-ha56f1ee_6\n",
            "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.3.0-h6de172a_19\n",
            "  lzo                conda-forge/linux-64::lzo-2.10-h516909a_1000\n",
            "  ncurses            conda-forge/linux-64::ncurses-6.2-h58526e2_4\n",
            "  nose               conda-forge/linux-64::nose-1.3.7-py27h8c360ce_1004\n",
            "  numpy              conda-forge/linux-64::numpy-1.16.5-py27h95a1406_0\n",
            "  openssl            conda-forge/linux-64::openssl-1.1.1k-h7f98852_0\n",
            "  pandas             conda-forge/linux-64::pandas-0.24.2-py27hb3f55d8_0\n",
            "  perl               conda-forge/linux-64::perl-5.32.1-0_h7f98852_perl5\n",
            "  pip                conda-forge/noarch::pip-20.1.1-pyh9f0ad1d_0\n",
            "  pybedtools         bioconda/linux-64::pybedtools-0.8.2-py27h6a42192_1\n",
            "  pysam              bioconda/linux-64::pysam-0.16.0.1-py27ha863e18_1\n",
            "  python             conda-forge/linux-64::python-2.7.15-h5a48372_1011_cpython\n",
            "  python-dateutil    conda-forge/noarch::python-dateutil-2.8.1-py_0\n",
            "  python-lzo         conda-forge/linux-64::python-lzo-1.12-py27h42e1302_1001\n",
            "  python_abi         conda-forge/linux-64::python_abi-2.7-1_cp27mu\n",
            "  pytz               conda-forge/noarch::pytz-2020.1-pyh9f0ad1d_0\n",
            "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
            "  samtools           bioconda/linux-64::samtools-1.10-h2e538c0_3\n",
            "  scipy              conda-forge/linux-64::scipy-1.2.1-py27h921218d_2\n",
            "  setuptools         conda-forge/linux-64::setuptools-44.0.0-py27_0\n",
            "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0\n",
            "  sqlite             conda-forge/linux-64::sqlite-3.35.5-h74cdb3f_0\n",
            "  tk                 conda-forge/linux-64::tk-8.6.10-h21135ba_1\n",
            "  wheel              conda-forge/noarch::wheel-0.36.2-pyhd3deb0d_0\n",
            "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
            "  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1010\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate happy-env\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - rtg-tools\n",
            "\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  ca-certificates    pkgs/main::ca-certificates-2020.10.14~ --> conda-forge::ca-certificates-2021.5.30-ha878542_0\n",
            "  certifi            pkgs/main/noarch::certifi-2020.6.20-p~ --> conda-forge/linux-64::certifi-2021.5.30-py38h578d9bd_0\n",
            "  conda               pkgs/main::conda-4.9.2-py38h06a4308_0 --> conda-forge::conda-4.10.1-py38h578d9bd_0\n",
            "  openssl                                 1.1.1h-h7b6447c_0 --> 1.1.1k-h27cfd23_0\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "Warning: 'defaults' already in 'channels' list, moving to the top\n",
            "Warning: 'bioconda' already in 'channels' list, moving to the top\n",
            "Warning: 'conda-forge' already in 'channels' list, moving to the top\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m4KjEpvKjBOG",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7b6fda60-d0bb-4489-fb34-53b9ce1de3fb"
      },
      "source": [
        "%%bash\n",
        "# run hap.py for benchmarking\n",
        "INPUT_DIR=\"/content/data\"\n",
        "OUTPUT_DIR=\"/content/clair3_ilmn_demo\"\n",
        "BASELINE_VCF_FILE_PATH=\"HG003_GRCh38_chr20_v4.2.1_benchmark.vcf.gz\"\n",
        "BASELINE_BED_FILE_PATH=\"HG003_GRCh38_chr20_v4.2.1_benchmark_noinconsistent.bed\"\n",
        "OUTPUT_VCF_FILE_PATH=\"merge_output.vcf.gz\"\n",
        "REF=\"GRCh38_chr20.fa\"\n",
        "CONTIGS=\"chr20\"\n",
        "START_POS=100000\n",
        "END_POS=300000\n",
        "THREADS=4\n",
        "source activate happy-env\n",
        "conda install -c bioconda rtg-tools -y \n",
        "hap.py \\\n",
        "    ${INPUT_DIR}/${BASELINE_VCF_FILE_PATH} \\\n",
        "    ${OUTPUT_DIR}/${OUTPUT_VCF_FILE_PATH} \\\n",
        "    -f \"${INPUT_DIR}/${BASELINE_BED_FILE_PATH}\" \\\n",
        "    -r \"${INPUT_DIR}/${REF}\" \\\n",
        "    -o \"${OUTPUT_DIR}/happy\" \\\n",
        "    -l ${CONTIGS}:${START_POS}-${END_POS} \\\n",
        "    --engine=vcfeval \\\n",
        "    --threads=\"${THREADS}\" \\\n",
        "    --pass-only"
      ],
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting package metadata (current_repodata.json): ...working... done\n",
            "Solving environment: ...working... done\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local/envs/happy-env\n",
            "\n",
            "  added / updated specs:\n",
            "    - rtg-tools\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  font-ttf-dejavu-s~ conda-forge/noarch::font-ttf-dejavu-sans-mono-2.37-hab24e00_0\n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.13.1-hba837de_1005\n",
            "  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1\n",
            "  icu                conda-forge/linux-64::icu-68.1-h58526e2_0\n",
            "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
            "  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2\n",
            "  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000\n",
            "  libxml2            conda-forge/linux-64::libxml2-2.9.12-h72842e0_0\n",
            "  openjdk            conda-forge/linux-64::openjdk-10.0.2-h14c3975_1015\n",
            "  rtg-tools          bioconda/noarch::rtg-tools-3.12.1-hdfd78af_0\n",
            "\n",
            "\n",
            "Preparing transaction: ...working... done\n",
            "Verifying transaction: ...working... done\n",
            "Executing transaction: ...working... done\n",
            "Hap.py \n",
            "Benchmarking Summary:\n",
            "  Type Filter  TRUTH.TOTAL  TRUTH.TP  TRUTH.FN  QUERY.TOTAL  QUERY.FP  QUERY.UNK  FP.gt  FP.al  METRIC.Recall  METRIC.Precision  METRIC.Frac_NA  METRIC.F1_Score  TRUTH.TOTAL.TiTv_ratio  QUERY.TOTAL.TiTv_ratio  TRUTH.TOTAL.het_hom_ratio  QUERY.TOTAL.het_hom_ratio\n",
            " INDEL    ALL           59        59         0          102         0         43      0      0       1.000000          1.000000        0.421569         1.000000                     NaN                     NaN                   4.090909                   4.210526\n",
            " INDEL   PASS           59        59         0          102         0         43      0      0       1.000000          1.000000        0.421569         1.000000                     NaN                     NaN                   4.090909                   4.210526\n",
            "   SNP    ALL          402       386        16          451         1         63      0      0       0.960199          0.997423        0.139690         0.978457                2.535088                2.205674                   4.647887                   4.294118\n",
            "   SNP   PASS          402       386        16          451         1         63      0      0       0.960199          0.997423        0.139690         0.978457                2.535088                2.205674                   4.647887                   4.294118\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "2021-06-16 12:18:13,295 WARNING  No reference file found at default locations. You can set the environment variable 'HGREF' or 'HG19' to point to a suitable Fasta file.\n",
            "[I] Total VCF records:         84252\n",
            "[I] Non-reference VCF records: 84252\n",
            "[I] Total VCF records:         888\n",
            "[I] Non-reference VCF records: 549\n",
            "2021-06-16 12:18:15,727 WARNING  starting at chr20:99999\n",
            "2021-06-16 12:18:15,911 WARNING  No calls for location chr20:100000-300000 in query!\n",
            "2021-06-16 12:18:15,911 WARNING  Creating template for vcfeval. You can speed this up by supplying a SDF template that corresponds to /content/data/GRCh38_chr20.fa\n"
          ],
          "name": "stderr"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "80EDfGqGwm2J"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}